### PR TITLE
Omnibus PR, including switch to explicit style differentiation

### LINKED ISSUE #221: Stop using implicit style differentiating
It seems the style used  here is being deprecated and  won't work with Flux 0.14: https://github.com/FluxML/MLJFlux.jl/blob/452c09d7dc13914f4057c661448bc310c9362d3d/src/core.jl#L37

---

**edit** After discussion below, I suggest we wait on

- [ ] https://github.com/FluxML/Optimisers.jl/pull/159

and refactor to use a optimiser-based solution to weight regularisation, which will avoid current limitations of explicit differentiation outlined in the discussion. Note, this will likely mean the reported `training_loss` must change, as it will no longer include the weight penalty. So this will be breaking.

### PR DESCRIPTION
**Pull request by ablaom**
This PR combines a number of changes, which for technical reasons could not be easily split up. The most important change, anticipating a Flux 0.15 breakage, is the switch to explicit differentiation; so this PR replaces #230. Shout out to @pat-alt for outlining a solution there. 

Closes #221.

To do:

- [x] Replace implicit style parameter updates with explicit style parameter updates, in
      line with planned Zygote/Flux deprecations.

- [x] Refactor code to use optimisers from Optimisers.jl with `setup/update` pattern in
      place of `update!` pattern. Also, rename private methods `train!` -> `train_epoch`
      and `fit!` -> `train` to reflect new non-mutating behaviour. This possibly breaks
      some "custom" models that have chosen to overload these technically private methods.

- [x] (**RNG changes.**) Change the default value of the model field `rng` from
      `Random.GLOBAL_RNG` to `Random.default_rng()`. Change the seeded RNG, obtained by
      specifying an integer value for `rng`, from `MersenneTwister` to `Xoshiro`.
  
- [x] Update the `Short` builder so that the `rng` argument of `build(::Short, rng, ...)`
      is passed on to the `Dropout` layer, as these layers now support this on a GPU, at
      least for `rng=Random.default_rng()`.

- [x] Change the implementation of L1/L2 regularization from explicit loss penalization to
      weight/sign decay (internally chained with the user-specified
      optimiser). **Breaking:** The losses reported in the history will no longer be
      penalized, because the penalty is not explicitly computed.

- [ ] Update documentation to reflect use of `Optimisers.jl` optimisers, instead of Flux.jl native optimisers. And on changes to the `rng` defaults. Waiting on #252

### CODE REVIEW

**Comment by ablaom** [comment]
@DilumAluthge Sorry to bother you, but I'm having trouble inspecting the BuildKite log here. It seems they are hidden from me, while previously they were not. Do I need to create a personal account to login and view them? Do I not have access as a FluxML member? What am I missing here?

**Comment by ablaom** [comment]
This PR is above average in complexity. This means a review is particularly important but it's also going to be more work than usual. @pat-alt Do you have any interest and time to review over the next 3 weeks, say? 

My apologies in advance for slightly messy commits. I temporarily lost access to a GPU for local testing and was shooting blind for a while.

File: test/mlj_model_interface.jl [hunk 1]
```diff
@@ -37,6 +37,48 @@ end
...
+        chain = MLJFlux.build(builder, StableRNG(123), 3, 1);
+        penalty = Penalizer(lambda, alpha); # defined in test_utils.jl
+        X, y = MLJFlux.collate(model, Xuser, yuser);
+        loss = model.loss;
+        n_batches = div(nobservations, batch_size)
+        optimiser_state = Optimisers.setup(optimiser, chain);
+        for i in 1:n_batches
+            batch_loss, gs = Flux.withgradient(chain) do m
+                yhat = m(X[i])
+                loss(yhat, y[i]) + sum(penalty, Optimisers.trainables(m))/n_batches
```
Comment (ablaom) [inline test/mlj_model_interface.jl:70]: @pat-alt Just FYI: Optimisers.jl added `trainables` which offers this solution to refactor along your original line. It superficially seems like `Flux.params`, but doesn't seem to suffer the same problems we were seeing (if you can remember back that far :wink:)

**Comment by ablaom** [comment]
@pat-alt Sorry to re-ping, but I'm not sure who else to ask for a review here. 
@tiemvanderdeure Would you consider reviewing?

If possible, hoping for a merge in the next 3 weeks. Even a cursory look, would be much appreciated!

**Comment by pat-alt** [comment]
Hi! I'll try to have a look as soon as I can (probably on the weekend or next week).

**Comment by tiemvanderdeure** [comment]
If it is helpful I can also give reviewing this PR a go. Probably won't have time the next few days but early next week should be feasible.

**Comment by ablaom** [comment]
Let's see if @pat-alt is able to finds some time.

File: src/builders.jl [hunk 2]
```diff
@@ -14,13 +14,12 @@
 abstract type Builder <: MLJModelInterface.MLJType end
 
 """
-    Linear(; σ=Flux.relu, rng=Random.GLOBAL_RNG)
```
Comment (pat-alt) [inline src/builders.jl:17]: Is it worth deprecating this?

File: src/core.jl [hunk 3]
```diff
@@ -24,38 +31,55 @@ Update the parameters of a Flux `chain`, where:
...
+        optimiser_state, chain = Optimisers.update(optimiser_state, chain, ∇)
     end
-    return training_loss / n_batches
+
+    return chain, optimiser_state, training_loss / n_batches
 end
 
 
 """
-    fit!(model::MLJFlux.MLJFluxModel, penalty, chain, optimiser, epochs, verbosity, X, y)
```
Comment (pat-alt) [inline src/core.jl:52]: I think this, too, might be worth deprecating. If I understand this correctly, existing extensions that overload `MLJFlux.fit!` like [here](https://github.com/JuliaTrustworthyAI/JointEnergyModels.jl/blob/724d6560bf27da48e4d60bfd5980661719153b09/src/mlj_flux.jl#L111) won't work anymore? As in, they should now be overloading the `train` function?

File: src/mlj_model_interface.jl [hunk 4]
```diff
@@ -43,7 +49,26 @@ end
 const ERR_BUILDER =
     "Builder does not appear to build an architecture compatible with supplied data. "
 
-true_rng(model) = model.rng isa Integer ? MersenneTwister(model.rng) : model.rng
+true_rng(model) = model.rng isa Integer ? Random.Xoshiro(model.rng) : model.rng
```
Comment (pat-alt) [inline src/mlj_model_interface.jl:52]: Just of curiosity, why this change?

File: src/mlj_model_interface.jl [hunk 5]
```diff
@@ -43,7 +49,26 @@ end
...
+
+# Models implement L1/L2 regularization by chaining the chosen optimiser with weight/sign
+# decay.  Note that the weight/sign decay must be scaled down by the number of batches to
+# ensure penalization over an epoch does not scale with the choice of batch size; see
+# https://github.com/FluxML/MLJFlux.jl/issues/213.
+
+function regularized_optimiser(model, nbatches)
```
Comment (pat-alt) [inline src/mlj_model_interface.jl:59]: So if I read this correctly, it is not possible to specify either L1 or L2? The regularized optimiser will always apply both?

File: test/mlj_model_interface.jl (hunk 1 above)
Comment (pat-alt) [reply test/mlj_model_interface.jl:70]: Thanks! 

So from a user-persective, I'm currently unclear about a few things (perhaps I've just not looked carefully enough!):

1. Is the `regularised_optimiser` method user-facing? Where/when is it used here in this specific unit test?
2. Would it make sense to ship the `Penalizer`?

**Review by pat-alt — COMMENTED** [review]
I mostly have a few comments about 1) deprecation and 2) clarification of how regularization is actually implemented on the user-end.

Otherwise, this looks good to me!

File: src/core.jl (hunk 3 above)
Comment (pat-alt) [reply src/core.jl:52]: So just to double-check @ablaom, `fit!` will become `train` and `train!` will become `train_epoch`?

**Comment by ablaom** [comment]
@pat-alt How are we doing? Happy with the changes?

**Review by pat-alt — APPROVED** [review]

**Comment by pat-alt** [comment]
> @pat-alt How are we doing? Happy with the changes?

Sorry, yes, I missed the thumbs up :) Thanks for clarifying!

**Comment by pat-alt** [comment]
[1bd58dd](https://github.com/FluxML/MLJFlux.jl/pull/251/commits/1bd58ddd9579e80f8402347241943964437cd6b0) adds deprecations for the `fit!` and `train!` methods. I think this is useful for developers who have used the old API in their own packages to extend MLJFlux, as we have done [here](https://github.com/JuliaTrustworthyAI/JointEnergyModels.jl/blob/724d6560bf27da48e4d60bfd5980661719153b09/src/mlj_flux.jl#L111), for example. Not sure who else has done something like this (and also unsure if this is the intended way to extend MLJFlux), but in any case adding these deprecations should help.

**Comment by ablaom** [comment]
Thanks @pat-alt  for your review. 🙏🏾
